pkg/utils/shuffle_test.go

```diff
@@ -0,0 +1,63 @@
+//
+// Copyright (c) 2018 The heketi Authors
+//
+// This file is licensed to you under your choice of the GNU Lesser
+// General Public License, version 3 or any later version (LGPLv3 or
+// later), or the GNU General Public License, version 2 (GPLv2), in all
+// cases as published by the Free Software Foundation.
+//
+
+package utils
+
+import (
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/heketi/tests"
+)
+
+func TestShuffle(t *testing.T) {
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+	a := []string{
+		"bobcat",
+		"lion",
+		"puma",
+		"cheetah",
+		"lynx",
+	}
+	a2 := make([]string, len(a))
+	copy(a2, a)
+	Shuffle(r, len(a2), func(i, j int) {
+		a2[i], a2[j] = a2[j], a2[i]
+	})
+	same := true
+	for i := 0; i < len(a); i++ {
+		same = same && (a[i] == a2[i])
+	}
+	tests.Assert(t, !same, a, a2)
+}
+
+func TestSeededShuffle(t *testing.T) {
+	a := []string{
+		"bobcat",
+		"lion",
+		"puma",
+		"cheetah",
+		"lynx",
+	}
+	a2 := make([]string, len(a))
+	dedup := map[string]int{}
+	copy(a2, a)
+	SeededShuffle(len(a2), func(i, j int) {
+		a2[i], a2[j] = a2[j], a2[i]
+	})
+	same := true
+	for i := 0; i < len(a); i++ {
+		dedup[a2[i]] += 1
+		same = same && (a[i] == a2[i])
+	}
+	tests.Assert(t, !same, a, a2)
```

Comment:
I expect this test to fail every 120th run on average, because that is the likelihood, that shuffling of 5 elements returns the identical sequence.

Reply:
Gosh, accelent point! This is just plain wrong. How can we test that the hosts are not always in the same order but can sometimes be?

Reply:
I would not test it, to be honest, else we are just retesting Go's implementation. I would probably only test it with a statically seeded case, so it's deterministic. Or repeat the test a couple of times? But this really sounds "over-tested" to me.

Reply:
Well, repeating makes it exponentially less likely that you have a "collision", so that would work. But as I said, it sounds like too much. All we actually want to test, is that shuffle is called. So I personally would seed the global PRNG with a fixed number before the test. And in production code I would make sure that we do proper seeding of it in the main function.